lib/Runtime/Provisioner/Provisioner.cpp

```diff
@@ -115,18 +142,31 @@ llvm::Error Provisioner::provision(DAGListTy &networks, Module &module,
         for (size_t i = 0, e = backends_.size(); i < e; i++) {
           if (backends_[i]->getBackendName() == nodeBackendName) {
             auto compiledOrErr = backends_[i]->compile(function, options);
+            // Check to see if an error was encountered while compiling.
             if (!compiledOrErr) {
+              // If and error occured, clean up provisioning state and return
+              // the error.
+              {
+                std::lock_guard<std::mutex> functionLock(functionsLock_);
+                cleanupProvision(localActiveNames);
```

Comment:
It looks to me as though you always grab the functionsLock_ before calling cleanupProvision, consider moving the lock into the method call to avoid missing it.